In [1]:
import pandas as pd


[Comparative study](https://www.researchgate.net/publication/327886491_A_Comparative_Study_of_Publicly_Available_Russian_Sentiment_Lexicons_7th_International_Conference_AINL_2018_St_Petersburg_Russia_October_17-19_2018_Proceedings)

|   **Lexicon name**    |**Year developed**|**Article description**|                                   **Dataset link**                                   |
|:---------------------:|:----------------:|:---------------------:|:------------------------------------------------------------------------------------:|
|    ProductSentiRus    |2012              |[Extraction of Russian Sentiment Lexicon for Product Meta-Domain](https://pdfs.semanticscholar.org/e23b/1430b3c4c4850db1336c5ba9c51c2084f29b.pdf)||
|        EmoLex         |2013              |[Crowdsourcing a Word-Emotion Association](https://arxiv.org/abs/1308.6297)|                [link](http://sentiment.nrc.ca/lexicons-for-research/)                |
| Chen-Skiena's Lexicon |2014         |[Building Sentiment Lexicons for All Major Languages](https://www.aclweb.org/anthology/P14-2063.pdf)| [link](https://sites.google.com/site/datascienceslab/projects/multilingualsentiment) |
|      LinisCrowd       |2016              |[An Opinion Word Lexicon and a Training Dataset for Russian Sentiment Analysis of Social Media](http://www.dialog-21.ru/media/3400/koltsovaoyuetal.pdf)|                         [link](http://www.linis-crowd.org/)                          |
|      RuSentiLex       |2017              |[Creating a General Russian Sentiment Lexicon](http://www.lrec-conf.org/proceedings/lrec2016/pdf/285_Paper.pdf)|       [.txt file](http://www.labinform.ru/pub/rusentilex/rusentilex_2017.txt)        |
|          MCF          |2017              |[Multilingual Connotation Frames: A Case Study on Social Media for Targeted Sentiment Analysis and Forecast.](aclanthology.org/P17-2073/)| [drive link](https://drive.google.com/file/d/1qxUM3TD-cFXzrGnr6yPwYeWZ9LNfspFr/view) |

# Chen-Skiena’s Lexicon

In [134]:
df_chen = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/parsed lexicons/ChenSkiena.csv', sep=';')

In [135]:
df_chen.polarity.unique()

array([-1,  1])

In [136]:
df_chen[df_chen['polarity']==1]

word  polarity
1648         как         1
1649         так         1
1650      первый         1
1651        типа         1
1652      именно         1
...          ...       ...
2909  улучшаться         1
2910  хорошенько         1
2911     хватить         1
2912  панорамный         1
2913    нежность         1

[1266 rows x 2 columns]

In [137]:
df_chen[df_chen['polarity']==-1]

word  polarity
0           время        -1
1              во        -1
2           около        -1
3             род        -1
4            член        -1
...           ...       ...
1643    неудачник        -1
1644   заморозить        -1
1645       травля        -1
1646     месячные        -1
1647  недомогание        -1

[1648 rows x 2 columns]

In [138]:
df_emolex = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/parsed lexicons/EmoLex.сsv', sep=';')

In [139]:
df_emolex

word  Positive  Negative
0              сзади         0         0
1              счеты         0         0
2      отказаться от         0         1
3        заброшенный         0         1
4         оставление         0         1
...              ...       ...       ...
14176           зона         0         0
14177        зоопарк         0         0
14178  зоологический         0         0
14179       зоология         0         0
14180            зум         0         0

[14181 rows x 3 columns]

In [140]:
df_emolex[df_emolex['Positive']==1]

word  Positive  Negative
7                  авва         1         0
20          способность         1         0
36       NO TRANSLATION         1         0
48           абсолютный         1         0
49     отпущение грехов         1         0
...                 ...       ...       ...
14162            моложе         1         0
14163          молодежь         1         0
14166            рвение         1         0
14168        ревностный         1         0
14173             цедра         1         0

[2312 rows x 3 columns]

In [141]:
df_emolex[df_emolex['Negative']==1]

word  Positive  Negative
2        отказаться от         0         1
3          заброшенный         0         1
4           оставление         0         1
13           отведение         0         1
14       отклоняющийся         0         1
...                ...       ...       ...
14143        тоскующий         1         1
14146           вопить         0         1
14148          желтеет         0         1
14149          тявкать         0         1
14174  застежка-молния         0         1

[3324 rows x 3 columns]

In [142]:
df_emolex[(df_emolex['Positive']==0)&(df_emolex['Negative']==0)]

word  Positive  Negative
0              сзади         0         0
1              счеты         0         0
5            утихать         0         0
6           снижение         0         0
8              аббат         0         0
...              ...       ...       ...
14176           зона         0         0
14177        зоопарк         0         0
14178  зоологический         0         0
14179       зоология         0         0
14180            зум         0         0

[8626 rows x 3 columns]

In [143]:
df_emolex[(df_emolex['Positive']==1)&(df_emolex['Negative']==1)]

word  Positive  Negative
61        изобилие         1         1
717    вооруженный         1         1
1017       бальзам         1         1
1365   хвастовство         1         1
1381     неистовый         1         1
...            ...       ...       ...
13477  развязывать         1         1
13510       в гору         1         1
13793        голос         1         1
13919          вес         1         1
14143    тоскующий         1         1

[81 rows x 3 columns]

# Linis crowd

In [116]:
df_linis = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/parsed lexicons/LinisCrowd.csv', sep=';')

In [117]:
df_linis

word  polarity
0          абажур        -1
1        абориген        -1
2           аборт        -1
3      абортивный         0
4      абортивный        -1
...           ...       ...
26759        яшин         0
26760        ящик         0
26761        ящик         0
26762        ящик         0
26763        ящик         0

[26764 rows x 2 columns]

In [118]:
df_linis.polarity.unique()

array([-1,  0, -2,  1,  2])

In [119]:
df_linis[df_linis['polarity']==-2]

word  polarity
61             авария        -2
79              аврал        -2
142    авторитарность        -2
177            агония        -2
191          агрессия        -2
...               ...       ...
26405      экстремист        -2
26407      экстремист        -2
26535     эротический        -2
26664              яд        -2
26735            ярмо        -2

[1136 rows x 2 columns]

In [120]:
df_linis[df_linis['polarity']==-1]

word  polarity
0          абажур        -1
1        абориген        -1
2           аборт        -1
4      абортивный        -1
16         абсурд        -1
...           ...       ...
26740      ярость        -1
26741      ярость        -1
26742      ярость        -1
26745        ярый        -1
26753       ясный        -1

[7026 rows x 2 columns]

In [121]:
df_linis_unique = df_linis.drop_duplicates(keep='first')

In [122]:
len(df_linis_unique)/len(df_linis)

0.4210880286952623

In [123]:
df_linis_unique[df_linis_unique['polarity']==-2]

word  polarity
61             авария        -2
79              аврал        -2
142    авторитарность        -2
177            агония        -2
191          агрессия        -2
...               ...       ...
26403      экстремизм        -2
26405      экстремист        -2
26535     эротический        -2
26664              яд        -2
26735            ярмо        -2

[747 rows x 2 columns]

In [124]:
df_linis_unique[df_linis_unique['polarity']==-1]

word  polarity
0          абажур        -1
1        абориген        -1
2           аборт        -1
4      абортивный        -1
16         абсурд        -1
...           ...       ...
26734        ярмо        -1
26737    яростный        -1
26740      ярость        -1
26745        ярый        -1
26753       ясный        -1

[3207 rows x 2 columns]

In [125]:
df_linis_unique[df_linis_unique['polarity']==0]

word  polarity
3       абортивный         0
7       абсолютный         0
10     абстрактный         0
21     абсурдность         0
29           абхаз         0
...            ...       ...
26748      ясность         0
26751        ясный         0
26754         яхта         0
26757         яшин         0
26760         ящик         0

[5563 rows x 2 columns]

In [126]:
df_linis_unique[df_linis_unique['polarity']==1]

word  polarity
92        австрийский         1
106              авто         1
125        автомобиль         1
139      авторитаризм         1
149         авторитет         1
...               ...       ...
26637  юмористический         1
26714          япония         1
26726           яркий         1
26746            ярый         1
26749         ясность         1

[1484 rows x 2 columns]

In [127]:
df_linis_unique[df_linis_unique['polarity']==2]

word  polarity
152     авторитетный         2
558       аристократ         2
668      атмосферный         2
952       башковитый         2
1261     безупречный         2
...              ...       ...
26366  экспрессивный         2
26380         экстаз         2
26575      эталонный         2
26577      этический         2
26752          ясный         2

[269 rows x 2 columns]

# ProductSentiRus

In [98]:
df_product = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/ProductSentiRus.csv', sep=';')

In [99]:
df_product

word  polarity
0         бесподобный     0.963
1           невнятный     0.953
2         отличнейший     0.935
3          обалденный     0.933
4             безумно     0.924
...               ...       ...
4995  присматриваться     0.145
4996         крикнуть     0.145
4997           богато     0.144
4998        скрипучий     0.144
4999     режиссерский     0.144

[5000 rows x 2 columns]

In [100]:
df_product.min()

word         аааа
polarity    0.144
dtype: object

In [101]:
df_product.max()

word        ясный
polarity    0.963
dtype: object

In [102]:
df_product.describe()

polarity
count  5000.000000
mean      0.300071
std       0.157558
min       0.144000
25%       0.186000
50%       0.244000
75%       0.364000
max       0.963000

In [103]:
df_product[df_product['polarity']<=df_product['polarity'].mean()]

word  polarity
1780             меню     0.300
1781     беспокоиться     0.300
1782       мультяшный     0.300
1783       переданный     0.300
1784      миниатюрный     0.300
...               ...       ...
4995  присматриваться     0.145
4996         крикнуть     0.145
4997           богато     0.144
4998        скрипучий     0.144
4999     режиссерский     0.144

[3220 rows x 2 columns]

# RuSentiLex

In [7]:
df_RuSentiLex = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/RuSentiLex.csv', sep=';')

In [8]:
df_RuSentiLex

word   pos        lemma sentiment   source  ambiguity rest
0            аборт  Noun        аборт  negative     fact        NaN  NaN
1       абортивный   Adj   абортивный  negative     fact        NaN  NaN
2      абракадабра  Noun  абракадабра  negative  opinion        NaN  NaN
3           абсурд  Noun       абсурд  negative  opinion        NaN  NaN
4      абсурдность  Noun  абсурдность  negative  opinion        NaN  NaN
...            ...   ...          ...       ...      ...        ...  ...
16049         ярый   Adj         ярый  positive  opinion  ПРЕДАННЫЙ  NaN
16050      ясность  Noun      ясность  positive  opinion        NaN  NaN
16051        ясный   Adj        ясный  positive  opinion        NaN  NaN
16052         ящур  Noun         ящур  negative     fact        NaN  NaN
16053      ящурный   Adj      ящурный  negative     fact        NaN  NaN

[16054 rows x 7 columns]

In [20]:
df_RuSentiLex.source.unique()

array(['fact', 'opinion', 'fact ', 'opinion ', 'opinion  ', 'feeling',
       'operator', 'negative', nan, 'feeling ',
       'opinion,"БЮРОКРАТИЗМ (ФОРМАЛИЗ,АНЦЕЛЯРЩИНА)"', 'fact В ТЕЗ EMPTY',
       'opinion    ', ' opinion', 'opinegativenion', ' fact', 'feeling  ',
       'ПРЕСНАЯ ВОДА', 'fact,"ПРЕСНАЯ ВОДА"', 'feeeling', 'operator ',
       'factg', 'opinon', 'fact В ТЕЗ empty',
       'opinion                     '], dtype=object)

Giving 0 to all facts, and 1 for feelings and opinions (lots of typos there though)

In [21]:
df_RuSentiLex['opinion_score'] = df_RuSentiLex['source'].apply(lambda x: 0 if 'fact' in str(x) else 1)

In [40]:
df_RuSentiLex['negative_sentiment_score'] = df_RuSentiLex['sentiment'].apply(lambda x: 1 if 'negative' in str(x) else 0)

In [42]:
df_RuSentiLex['negative_opinion_score'] = df_RuSentiLex['opinion_score'] + df_RuSentiLex['negative_sentiment_score']

In [43]:
df_RuSentiLex

word   pos        lemma sentiment   source  ambiguity rest  \
0            аборт  Noun        аборт  negative     fact        NaN  NaN   
1       абортивный   Adj   абортивный  negative     fact        NaN  NaN   
2      абракадабра  Noun  абракадабра  negative  opinion        NaN  NaN   
3           абсурд  Noun       абсурд  negative  opinion        NaN  NaN   
4      абсурдность  Noun  абсурдность  negative  opinion        NaN  NaN   
...            ...   ...          ...       ...      ...        ...  ...   
16049         ярый   Adj         ярый  positive  opinion  ПРЕДАННЫЙ  NaN   
16050      ясность  Noun      ясность  positive  opinion        NaN  NaN   
16051        ясный   Adj        ясный  positive  opinion        NaN  NaN   
16052         ящур  Noun         ящур  negative     fact        NaN  NaN   
16053      ящурный   Adj      ящурный  negative     fact        NaN  NaN   

       opinion_score  negative_sentiment_score  negative_opinion_score  
0                  0                         1                       1  
1                  0                         1                       1  
2                  1                         1                       2  
3                  1                         1                       2  
4                  1                         1                       2  
...              ...                       ...                     ...  
16049              1                         0                       1  
16050              1                         0                       1  
16051              1                         0                       1  
16052              0                         1                       1  
16053              0                         1                       1  

[16054 rows x 10 columns]

# Multilingual connotation frames

In [147]:
with open('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/MCF/multilingual/multilingual_alignments/ru/ru_dict.txt', 'r') as file:
    data = file.readlines()

In [148]:
clean_data = []
for line in data:
    line = line.strip()
    values = line.split('\t')
    clean_data.append(values)

In [149]:
clean_data

[['ЭОД',
  'authorize,0.10',
  'head,0.10',
  'finish,0.10',
  'tap,0.05',
  'show,0.14',
  'process,0.05',
  'transfer,0.10',
  'make,0.14',
  'reiterate,0.05',
  'place,0.05',
  'take,0.05',
  'possess,0.05'],
 ['Лихтенштейном',
  'desire,0.04',
  'exceed,0.04',
  'motivate,0.04',
  'relate,0.04',
  'support,0.07',
  'originate,0.04',
  'list,0.07',
  'focus,0.04',
  'see,0.07',
  'intercept,0.04',
  'choose,0.04',
  'have,0.37',
  'report,0.07',
  'approach,0.04'],
 ['огромного',
  'major,0.07',
  'drop,0.01',
  'stream,0.02',
  'supply,0.01',
  'overwhelm,0.16',
  'render,0.02',
  'exceed,0.01',
  'impair,0.01',
  'fashion,0.01',
  'fail,0.01',
  'putt,0.01',
  'rely,0.01',
  'accomplish,0.01',
  'group,0.01',
  'deal,0.11',
  'support,0.01',
  'amount,0.09',
  'mark,0.01',
  'flood,0.01',
  'score,0.02',
  'astonish,0.01',
  'override,0.02',
  'daunt,0.01',
  'document,0.01',
  'return,0.01',
  'bear,0.01',
  'water,0.01',
  'explore,0.01',
  'capitalize,0.01',
  'expand,0.01',
  

In [150]:
from itertools import zip_longest
import numpy as np
num_columns = len(max(clean_data, key=len))

In [151]:
new_data = list(zip_longest(*clean_data, fillvalue=None))

In [152]:
df_mcc_alignments = pd.DataFrame(new_data).T

In [153]:
df_mcc_alignments.columns = [f'Column {i+1}' for i in range(num_columns)]

In [154]:
df_mcc_alignments[~df_mcc_alignments['Column 50'].isna()]

Column 1          Column 2        Column 3       Column 4  \
619              раннее          set,0.01    foresee,0.01    supply,0.02   
747     компьютеризации    represent,0.01        set,0.01  consider,0.01   
1115            встанут        enjoy,0.01      major,0.01      work,0.01   
3296       секторальные     conclude,0.02  attribute,0.01     trade,0.02   
5201       гармонизации  consolidate,0.01        set,0.05    burden,0.01   
...                 ...               ...             ...            ...   
220609            целым     organize,0.01      elect,0.01    stream,0.01   
221285   предполагающая       emerge,0.01      elect,0.01    entail,0.02   
221687         .Факторы     organize,0.01      force,0.01    regret,0.01   
223088     объективными        claim,0.01       hunt,0.01     force,0.01   
223558          БУДУЩЕЙ      comment,0.01      force,0.01      deal,0.01   

               Column 5         Column 6         Column 7         Column 8  \
619     strengthen,0.02      revert,0.01        side,0.01      number,0.03   
747           show,0.01     process,0.09        back,0.01      detail,0.01   
1115          have,0.02       bring,0.01         set,0.01     embrace,0.01   
3296    constitute,0.01       phase,0.01     exploit,0.01     measure,0.01   
5201        manage,0.01         tax,0.01      direct,0.03    identify,0.01   
...                 ...              ...              ...              ...   
220609          do,0.02      attend,0.01      manage,0.01     commend,0.01   
221285     purport,0.01     mandate,0.01     involve,0.17       bring,0.01   
221687    campaign,0.01    sentence,0.01        lack,0.01        sign,0.01   
223088     prevent,0.03      relate,0.01        mind,0.01  constitute,0.01   
223558       point,0.04  distribute,0.03  articulate,0.01       carry,0.01   

               Column 9      Column 10  ...   Column 62      Column 63  \
619        involve,0.01  organize,0.01  ...        None           None   
747         assess,0.01  organize,0.01  ...        None           None   
1115     undertake,0.01     arise,0.05  ...        None           None   
3296          need,0.01     close,0.01  ...        None           None   
5201      restrict,0.01     bring,0.09  ...        None           None   
...                 ...            ...  ...         ...            ...   
220609        date,0.01  identify,0.01  ...  agree,0.01  comprise,0.01   
221285       imply,0.05   embrace,0.01  ...        None           None   
221687        down,0.03     shape,0.01  ...        None           None   
223088       imply,0.01     sleep,0.01  ...        None           None   
223558  scrutinize,0.01     exist,0.01  ...        None           None   

       Column 64 Column 65 Column 66 Column 67 Column 68 Column 69 Column 70  \
619         None      None      None      None      None      None      None   
747         None      None      None      None      None      None      None   
1115        None      None      None      None      None      None      None   
3296        None      None      None      None      None      None      None   
5201        None      None      None      None      None      None      None   
...          ...       ...       ...       ...       ...       ...       ...   
220609      None      None      None      None      None      None      None   
221285      None      None      None      None      None      None      None   
221687      None      None      None      None      None      None      None   
223088      None      None      None      None      None      None      None   
223558      None      None      None      None      None      None      None   

       Column 71  
619         None  
747         None  
1115        None  
3296        None  
5201        None  
...          ...  
220609      None  
221285      None  
221687      None  
223088      None  
223558      None  

[203 rows x 71 columns]

In [155]:
df_mcf_conn = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/lexicons/MCF/multilingual/multilingual_connotation/ru_cf.txt', sep='\t', decimal='.')

In [158]:
len(df_mcf_conn)

212973

In [156]:
df_mcf_conn.tail(20)

ru Word English verb  P(w->t)  P(w->a)  P(a->t)  E(t)  E(a)  \
212953      натравливает           be        0        1        0     0     0   
212954              льше        total        0        0        0     0     0   
212955   ремилитаризацию      conduct        0        0        1     1     0   
212956            рожала      control        1        0        1     0     1   
212957       измышлениях       launch        1        1        1     1     1   
212958            вводам           go        0        0        0     0     0   
212959           отрубов          cut        0        0        0    -1     0   
212960       завершаются     conclude        1        1        1     1     0   
212961        улучшиться      improve        1        1        1     1     0   
212962  экспортировалась       export        0        0        1     0     1   
212963        проявившие      express       -1        0        0     0     1   
212964      кодироваться           be        0        1        0     0     0   
212965         достигнув        reach        0        1        0     0     0   
212966           Шымкент           be        0        1        0     0     0   
212967   водопроницаемая        water        0       -1        0    -1     0   
212968       подогревают         fuel        0        0        0     1     1   
212969          вечернем         work        1        1        1     1     0   
212970           сбитого        shoot        0        0        0    -1     0   
212971          вечернее           be        0        1        0     0     0   
212972          ночующие        sleep        0        1        0     0     1   

        V(t)  V(a)  S(t)  S(a)  
212953     0     1     0     1  
212954     1     1     0     0  
212955     1     1     0     1  
212956     1     1     0     1  
212957     1     1     0     1  
212958     0     1     0     0  
212959     0     1     0     0  
212960     1     1     0     1  
212961     1     1     0     1  
212962     1     1     0     1  
212963     0     1     0     0  
212964     0     1     0     1  
212965     1     1     0     0  
212966     0     1     0     1  
212967     1     1     0     1  
212968     1     1     0     1  
212969     1     1     0     1  
212970    -1     1    -1     0  
212971     0     1     0     1  
212972     1     1     0     1

In [157]:
df_mcc_alignments.tail(20)

Column 1          Column 2         Column 3         Column 4  \
225128       писаной        write,0.62     station,0.06       radio,0.06   
225129    захватывая  consolidate,0.04     capture,0.08          be,0.08   
225130       125-129     increase,0.20      relate,0.20     require,0.20   
225131        плодам        enjoy,0.09      afford,0.05       share,0.09   
225132   формируется        shape,0.04         set,0.01        draw,0.01   
225133      подобную     prohibit,0.02       force,0.02    conclude,0.02   
225134       Исабаль          map,0.14         set,0.29     attempt,0.14   
225135    набирающей  consolidate,0.03        rise,0.03       widen,0.03   
225136    набирающее   strengthen,0.25      signal,0.12      gather,0.25   
225137  ВЕРТИКАЛЬНАЯ      perform,0.30     require,0.10  contribute,0.10   
225138         КБХПР       assess,0.17      review,0.17        land,0.17   
225139     учреждает          set,0.04  constitute,0.07   establish,0.74   
225140   долговечным      sustain,0.08        last,0.08      endure,0.38   
225141  организующих        count,0.02    organize,0.48      handle,0.02   
225142        поводу       relate,0.03        lack,0.01     concern,0.09   
225143       писаном        write,0.43       print,0.14      suffer,0.14   
225144        поводы      respond,0.03   formulate,0.03        give,0.03   
225145       писаное           do,0.13     wrestle,0.07        deal,0.07   
225146     ненасилия        begin,0.05      engage,0.10        form,0.02   
225147   запутыванию       evolve,0.10  complicate,0.10     suggest,0.10   

               Column 5        Column 6         Column 7         Column 8  \
225128        have,0.12      press,0.06  understand,0.03   underline,0.03   
225129       seize,0.12     assume,0.04        ruin,0.04  confiscate,0.04   
225130   implement,0.20    propose,0.20             None             None   
225131     benefit,0.59    produce,0.09       place,0.05             None   
225132  constitute,0.02         be,0.46      emerge,0.09   establish,0.02   
225133     respect,0.02       seek,0.04     concern,0.02        form,0.03   
225134       level,0.14     create,0.14        grow,0.14             None   
225135     process,0.03     gather,0.07       grind,0.03    increase,0.17   
225136      mature,0.12       come,0.12        grow,0.12             None   
225137        beam,0.10      range,0.10         can,0.30             None   
225138       allow,0.17  determine,0.33             None             None   
225139      create,0.01         be,0.07             None             None   
225140        mark,0.46            None             None             None   
225141        lead,0.02    provide,0.13      create,0.02     procure,0.04   
225142     respect,0.08    subject,0.02       issue,0.03      regard,0.62   
225143     embrace,0.14      award,0.14             None             None   
225144     involve,0.03     design,0.03        want,0.03     concern,0.05   
225145        rank,0.07      write,0.53       exist,0.07       treat,0.07   
225146    practice,0.12    disable,0.05      amount,0.02      foster,0.56   
225147     confuse,0.10       tend,0.30     obscure,0.20  facilitate,0.10   

               Column 9       Column 10  ... Column 62 Column 63 Column 64  \
225128     subject,0.03            None  ...      None      None      None   
225129     involve,0.04       link,0.08  ...      None      None      None   
225130             None            None  ...      None      None      None   
225131             None            None  ...      None      None      None   
225132      create,0.02       form,0.06  ...      None      None      None   
225133  commission,0.02     detail,0.05  ...      None      None      None   
225134             None            None  ...      None      None      None   
225135     benefit,0.03       gain,0.10  ...      None      None      None   
225136             None            None  ...      None      None      None   
225137           

[Diachronic analysis](https://aclanthology.org/W19-4725.pdf)

In [1]:
from collocation_extraction import collect_verb_obl_obj, collect_core, collect_nmod, collect_amod, collect_comp, collect_appos, show_dependency
from text_preprocessing import preprocess_df
import pandas as pd
import spacy

In [2]:
df_labels = pd.read_csv('/Users/katerynaburovova/PycharmProjects/dehumanization/annotation/final_labels.csv', index_col=0)

In [3]:
df_labels = df_labels[df_labels['Dehumanization']!='не можу визначитись з правильною відповіддю'].copy()

In [4]:
df_labels['label'] = df_labels['Dehumanization'].apply(lambda x: 0 if x=='ні' else 1)

In [51]:
import spacy
# from spacy.lang.ru.examples import sentences
nlp = spacy.load('ru_core_news_md')
def lemmatize_spacy(text):
    doc = nlp(text)
    result = [token.lemma_ for token in doc]
    return result

In [52]:
%%time
df_labels['lemmas'] = df_labels['text'].apply(lambda x: lemmatize_spacy(x))

CPU times: user 29.5 s, sys: 244 ms, total: 29.7 s
Wall time: 30 s


In [32]:
#creating opinion score dict
opinion_dict_rusentilex = dict(zip(df_RuSentiLex.iloc[:, 0], df_RuSentiLex.iloc[:, 7]))

In [53]:
def opinion_score_from_text(text, sentiment_dict):
    words = text.split()
    score = 0

    for word in words:
        if word in sentiment_dict:
            score += sentiment_dict[word]

    return score


In [54]:
def opinion_score_from_lemmas(list, sentiment_dict):
    score = 0

    for word in list:
        if word in sentiment_dict:
            score += sentiment_dict[word]

    return score


In [33]:
opinion_dict_rusentilex['абракадабра']

1

In [55]:
df_labels['opinion_score'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, opinion_dict_rusentilex))

In [57]:
import numpy as np

pearson_correlation = df_labels["opinion_score"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

Pearson correlation coefficient: 0.06896494852519809


In [58]:
spearman_correlation = df_labels["opinion_score"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Spearman correlation coefficient: 0.08141875590397507


As one variable increases, the other tends to increase slightly as well, but the relationship is not strong.

In [60]:
negative_dict_rusentilex = dict(zip(df_RuSentiLex.iloc[:, 0], df_RuSentiLex.iloc[:, 8]))

In [61]:
negative_op_dict_rusentilex = dict(zip(df_RuSentiLex.iloc[:, 0], df_RuSentiLex.iloc[:, 9]))

In [62]:
df_labels['negative_emotion_score'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, negative_dict_rusentilex))
df_labels['negative_opinion_score'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, negative_op_dict_rusentilex))

In [63]:
pearson_correlation = df_labels["negative_emotion_score"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["negative_emotion_score"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.16588910696147194
Spearman correlation coefficient: 0.20957913960148952


In [64]:
pearson_correlation = df_labels["negative_opinion_score"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["negative_opinion_score"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.13121852288771577
Spearman correlation coefficient: 0.16701545979788543


The relationship is a bit better described by a non-linear, monotonic function rather than a linear function, and for negative emotion only it's stronger that for negative emotion and opinion combined.

A value of 0.20957913960148952 is relatively low, suggesting that the relationship between the two variables is not very strong. However, it is still positive, which indicates that some degree of association exists between the variables.

In [78]:
from collocation_extraction import collect_verb_obl_obj, collect_core, collect_nmod, collect_amod, collect_comp, collect_appos, show_dependency, lst_to_str

In [81]:
df_labels['core_noun_verb'] = df_labels['text'].apply(lambda x: lst_to_str(collect_core(x)))
df_labels['verb_obl_obj'] = df_labels['text'].apply(lambda x: lst_to_str(collect_verb_obl_obj(x)))
df_labels['nmod'] = df_labels['text'].apply(lambda x: lst_to_str(collect_nmod(x)))
df_labels['amod'] = df_labels['text'].apply(lambda x: lst_to_str(collect_amod(x)))
df_labels['comp'] = df_labels['text'].apply(lambda x: lst_to_str(collect_comp(x)))

In [84]:
df_labels['core_noun_verb_lemmas'] = df_labels['core_noun_verb'].apply(lambda x: lemmatize_spacy(x))
df_labels['verb_obl_obj_lemmas'] = df_labels['verb_obl_obj'].apply(lambda x: lemmatize_spacy(x))
df_labels['nmod_lemmas'] = df_labels['nmod'].apply(lambda x: lemmatize_spacy(x))
df_labels['amod_lemmas'] = df_labels['amod'].apply(lambda x: lemmatize_spacy(x))
df_labels['comp_lemmas'] = df_labels['comp'].apply(lambda x: lemmatize_spacy(x))

In [80]:
df_labels

Emotion Dehumanization Mention   External ID  \
1219  так, присутня негативна            так     так     row_0.txt   
1218   ні, оцінка не присутня             ні      ні     row_1.txt   
1591  так, присутня негативна            так     так    row_10.txt   
1198  так, присутня негативна            так     так   row_100.txt   
3247   ні, оцінка не присутня             ні     так  row_1000.txt   
...                       ...            ...     ...           ...   
3613   ні, оцінка не присутня             ні     так   row_995.txt   
3612  так, присутня негативна             ні     так   row_996.txt   
4121  так, присутня негативна             ні     так   row_997.txt   
4120   ні, оцінка не присутня            так     так   row_998.txt   
3248   ні, оцінка не присутня             ні      ні   row_999.txt   

                        Created By  \
1219      snizannabotvin@gmail.com   
1218      snizannabotvin@gmail.com   
1591      snizannabotvin@gmail.com   
1198      snizannabotvin@gmail.com   
3247        tutovadesign@gmail.com   
...                            ...   
3613        tutovadesign@gmail.com   
3612        tutovadesign@gmail.com   
4121  yevhen.marchenko91@gmail.com   
4120  yevhen.marchenko91@gmail.com   
3248        tutovadesign@gmail.com   

                                                   text  rating  label  \
1219  Всвязи с этим немного поправлю коллег ⤵️  "Они...       4      1   
1218  Литературный критик Галина Юзефович о новом ро...       4      0   
1591  Почему на базах неонацистов стоят языческие ис...       4      1   
1198  Группа добровольцев-медиков из Чеченской Респу...       4      1   
3247  ВСУшники, переходите на сторону добра, у нас т...       5      0   
...                                                 ...     ...    ...   
3613  Утренний брифинг Минобороны России:  ▪️ россий...       5      0   
3612  И понеслась мазепинщино-петлюровщино-бандеровщ...       5      0   
4121  Наш соратник по русскому движению Алексей Сели...       3      0   
4120  Хорошее видео от 4 бригады НМ ЛНР https://t.me...       3      1   
3248  Замоскворецкий районный суд Москвы отклонил ис...       5      0   

      opinion_score  negative_emotion_score  negative_opinion_score  \
1219              3                       2                       5   
1218              2                       0                       2   
1591              0                       1                       1   
1198              4                       2                       6   
3247              2                       0                       2   
...             ...                     ...                     ...   
3613              2                       4                       6   
3612              0                       0                       0   
4121              1                       0                       1   
4120              2                       1                       3   
3248              0                       1                       1   

                                                 lemmas  \
1219  [всвязи, с, это, немного, поправить, коллега, ...   
1218  [литературный, критик, галина, юзефович, о, но...   
1591  [почему, на, база, неонацист, стоять, язычески...   
1198  [группа, доброволец, -, медик, из, чеченский, ...   
3247  [всушники, ,, переходить, на, сторона, добро, ...   
...                                                 ...   
3613  [утренний, брифинг, минобороны, россия, :,  , ...   
3612  [и, понестись, мазепинщино, -, петлюровщино, -...   
4121  [наш, соратник, по, русский, движение, алексей...   
4120  [хороший, видео, от, 4, бригада, нм, лнр, http...   
3248  [замоскворецкий, районный, суд, москва, отклон...   

      negative_opinion_score_mult  \
1219                            5   
1218                            2   
1591                            1   
1198                            6   
3247                            2   
...                           ...   
3613          

In [93]:
df_labels.columns[18:]

Index(['core_noun_verb_lemmas', 'verb_obl_obj_lemmas', 'nmod_lemmas',
       'amod_lemmas', 'comp_lemmas'],
      dtype='object')

## Trying the collocations

In [94]:
collocation_columns = df_labels.columns[18:]
collocation_columns

Index(['core_noun_verb_lemmas', 'verb_obl_obj_lemmas', 'nmod_lemmas',
       'amod_lemmas', 'comp_lemmas'],
      dtype='object')

In [97]:
for column in collocation_columns:
    current_col_name = column + '_score'
    df_labels[current_col_name] = df_labels[column].apply(lambda x: opinion_score_from_lemmas(x, negative_dict_rusentilex))

    pearson_correlation = df_labels[current_col_name].corr(df_labels["label"], method="pearson")

    spearman_correlation = df_labels[current_col_name].corr(df_labels["label"], method="spearman")

    print(f'\n For column {column} the Pearson coef is {pearson_correlation} and Spearman coef is {spearman_correlation}')


 For column core_noun_verb_lemmas the Pearson coef is 0.08441653965818131 and Spearman coef is 0.11040692451113025

 For column verb_obl_obj_lemmas the Pearson coef is 0.10573957138156906 and Spearman coef is 0.13038628074912484

 For column nmod_lemmas the Pearson coef is 0.08212894400918958 and Spearman coef is 0.11346956239263108

 For column amod_lemmas the Pearson coef is 0.12713745218542027 and Spearman coef is 0.1487866502012851

 For column comp_lemmas the Pearson coef is nan and Spearman coef is nan


/Users/katerynaburovova/PycharmProjects/dehumanization/lib/python3.10/site-packages/scipy/stats/_stats_py.py:4921: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.
  warnings.warn(stats.ConstantInputWarning(warn_msg))


## Product_senti_rus

In [108]:
polarity_dict_product_senti_rus = dict(zip(df_product.iloc[:, 0], df_product.iloc[:, 1]))

In [110]:
polarity_dict_product_senti_rus['бесподобный']

0.963

In [111]:
df_labels['polarity_score_rod'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, polarity_dict_product_senti_rus))

In [112]:
pearson_correlation = df_labels["polarity_score_rod"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["polarity_score_rod"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.06205632028788792
Spearman correlation coefficient: 0.0683566575149784


## Linis crowd

In [128]:
polarity_dict_linis = dict(zip(df_linis.iloc[:, 0], df_linis.iloc[:, 1]))

In [130]:
df_labels['polarity_score_linis'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, polarity_dict_linis))

In [131]:
pearson_correlation = df_labels["polarity_score_linis"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["polarity_score_linis"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: -0.07257302614977595
Spearman correlation coefficient: -0.09172859355596684


## Chen-Skiena’s Lexicon

In [145]:
polarity_dict_chen = dict(zip(df_chen.iloc[:, 0], df_chen.iloc[:, 1]))
df_labels['polarity_score_chen'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, polarity_dict_chen))
pearson_correlation = df_labels["polarity_score_chen"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["polarity_score_chen"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: -0.024412514693019676
Spearman correlation coefficient: -0.027407066980927958


## Emolex

In [146]:
polarity_dict_emolex = dict(zip(df_emolex.iloc[:, 0], df_emolex.iloc[:, 2]))
df_labels['polarity_score_emolex'] = df_labels['lemmas'].apply(lambda x: opinion_score_from_lemmas(x, polarity_dict_emolex))
pearson_correlation = df_labels["polarity_score_emolex"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["polarity_score_emolex"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.08419974967156509
Spearman correlation coefficient: 0.09171169160415636


## Multilingual connotation frames

A framework for encoding predicate-specific connotative relationships implied by a predicate towards its arguments - unique lexical resource to enable targeted sentiment analysis.

In [160]:
df_mcf_conn

ru Word English verb  P(w->t)  P(w->a)  P(a->t)  E(t)  \
0                  фор-мах         form        1        1        1     1   
1             добавленными          add        1        1        1     1   
2              выполнялось           be        0        1        0     0   
3                 девятого        adopt        1        1        1     1   
4       правосубъектностью          own        0        1        1     1   
...                    ...          ...      ...      ...      ...   ...   
212968         подогревают         fuel        0        0        0     1   
212969            вечернем         work        1        1        1     1   
212970             сбитого        shoot        0        0        0    -1   
212971            вечернее           be        0        1        0     0   
212972            ночующие        sleep        0        1        0     0   

        E(a)  V(t)  V(a)  S(t)  S(a)  
0          1     1     1     0     1  
1          1     0     1     0     1  
2          0     0     1     0     1  
3          1     1     1     0     1  
4          1     1     1     0     1  
...      ...   ...   ...   ...   ...  
212968     1     1     1     0     1  
212969     0     1     1     0     1  
212970     0    -1     1    -1     0  
212971     0     0     1     0     1  
212972     1     1     1     0     1  

[212973 rows x 11 columns]

ru Word: The Russian word.
English verb: The English translation of the Russian word.
P(w->t): Probability that the word has a positive effect on its theme (object).
P(w->a): Probability that the word has a positive effect on its agent (subject).
P(a->t): Probability that the agent (subject) has a positive effect on the theme (object).
E(t): Expected value of the effect of the word on the theme (object).
E(a): Expected value of the effect of the word on the agent (subject).
V(t): Variance of the effect of the word on the theme (object).
V(a): Variance of the effect of the word on the agent (subject).
S(t): Skewness of the effect of the word on the theme (object).
S(a): Skewness of the effect of the word on the agent (subject).

In [164]:
%%time
df_mcf_conn['ru_Word_lemmatized'] = df_mcf_conn['ru Word'].apply(lambda x: lemmatize_spacy(str(x)))

CPU times: user 9min 50s, sys: 7.01 s, total: 9min 57s
Wall time: 10min 6s


In [171]:
df_mcf_conn['ru_Word_lemmatized_str'] = df_mcf_conn['ru_Word_lemmatized'].apply(lambda x: ' '.join(x))

In [175]:
def create_sentiment_dict(df_multicf):
    sentiment_dict = {}

    for index, row in df_multicf.iterrows():
        lemma = row['ru_Word_lemmatized_str']
        sentiment_dict[lemma] = {
            'P(w->t)': row['P(w->t)'],
            'P(w->a)': row['P(w->a)'],
            'P(a->t)': row['P(a->t)'],
        }

    return sentiment_dict

In [176]:
sentiment_dict_mcf_conn = create_sentiment_dict(df_mcf_conn)

In [177]:
def calculate_sentiment_scores(df, sentiment_dict, text_column):
    sentiment_scores = ['P(w->t)', 'P(w->a)', 'P(a->t)']
    for score_type in sentiment_scores:
        df[score_type] = df[text_column].apply(lambda text: calculate_score(text, sentiment_dict, score_type))

    return df

def calculate_score(lemmas, sentiment_dict, score_type):
    score = 0
    for lemma in lemmas:
        if lemma in sentiment_dict:
            score += sentiment_dict[lemma][score_type]

    return score

In [179]:
df_labels_with_scores = calculate_sentiment_scores(df_labels, sentiment_dict_mcf_conn, 'lemmas')

In [181]:
pearson_correlation = df_labels["P(w->t)"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["P(w->t)"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.031441116684933004
Spearman correlation coefficient: 0.038563021120012954


In [182]:
pearson_correlation = df_labels["P(w->a)"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["P(w->a)"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.03922600860881998
Spearman correlation coefficient: 0.05622414138632885


In [183]:
pearson_correlation = df_labels["P(a->t)"].corr(df_labels["label"], method="pearson")
print("Pearson correlation coefficient:", pearson_correlation)

spearman_correlation = df_labels["P(a->t)"].corr(df_labels["label"], method="spearman")
print("Spearman correlation coefficient:", spearman_correlation)

Pearson correlation coefficient: 0.027816119538881908
Spearman correlation coefficient: 0.04076976556056823
